backend/api/CourseApi/index.js

```diff
@@ -16,10 +16,20 @@ router.get('/public', catchAsync(async (request, response) => {
     offset: (pageNumber - 1) * pageSize,
     courseCategoryId: request.query.courseCategoryId
   });
-  const amountOfAllCourses = await CourseModel.select.countAllPublic({
-    courseCategoryId: request.query.courseCategoryId
-  });
 
+  var amountOfAllCourses;
+  var subquery = knex('problem').count('id').whereRaw('problem.course_id = course.id');
+  if(request.query.courseCategoryId) {
+    amountOfAllCourses = await knex('course').count('id', {as: 'amount_of_public_courses'})
+    .where('if_public', 'true')
+    .andWhere(2, "<=", subquery)
+    .andWhere('course_category_id', request.query.courseCategoryId);
+  } else {
+    amountOfAllCourses = await knex('course').count('id', {as: 'amount_of_public_courses'})
+    .where('if_public', 'true')
+    .andWhere(2, "<=", subquery); 
+  }
+  
```

Comment:
Looks nice!
We can improve it by factoring out this piece of code:

```
knex('course').count('id', {as: 'amount_of_public_courses'})
    .where('if_public', 'true')
    .andWhere(2, "<=", subquery); 
```

E.g. we can do (please check if this actually works though :-)):

```
const courseCategoryId = request.query.courseCategoryId;
let amountPromise = knex('course').count('id', {as: 'amount_of_public_courses'})
    .where('if_public', 'true')
    .andWhere(2, "<=", subquery); 

if (courseCategoryId) {
  amountPromise = amountPromise.andWhere('course_category_id', courseCategoryId)
}

const amountOfAllCourses = await amountPromise;
```

Reply:
I will check and revert. Also, will change the 'allPublic' so that we can completely get rid of older syntax and then will commit the changes. Thank you for reviewing.

Reply:
This is done in the new commit.